Fire event when Item Entity Expires

DrZoddiak opened the pull request:
Should resolve 1/2 of  #1584

Full disclosure I had no idea what I'm doing.
I just tinkered with it until it worked (with some assistance in discord)

It does fire an event every time an item entity expires. 
My test code: https://gist.github.com/DrZoddiak/6e542f070520e11193b8e2fd3c6f64f3

DrZoddiak committed:
Fire event when Item Entity Expires

Should resolve 1/2 of  #1584

dualspiral commented on src/main/java/org/spongepowered/common/mixin/core/entity/item/EntityItemMixin.java:
```diff
@@ -192,6 +192,22 @@
+            ordinal = 1, 
+            target = "Lnet/minecraft/entity/item/EntityItem;setDead()V")
+    )
+    private void fireExpireEntityEventTargetItem (final CallbackInfo ci) {
```
I hate to be the one to complain about formatting...

This method name needs to be prefixed with `spongeImpl$` to be consistent with the class (though, @gabizou, [I thought the policy was to prefix `impl$`, not `spongeImpl$`](https://github.com/SpongePowered/SpongeCommon/wiki/Implementation-Policy:-Bridges-and-Mixins#rules-3))? 

Also, very nitpicky, but while you're there, can you remove the space between the method name and parameters, and the space after the method name?

dualspiral reviewed (commented):
I've not tested this, but I do have a formatting comment.

A glance says it looks fine though, it's simply firing an event when an item decays. Nothing particularly special about that really...

DrZoddiak committed:
Add spongeImpl & fix formatting issues

gabizou commented on src/main/java/org/spongepowered/common/mixin/core/entity/item/EntityItemMixin.java:
```diff
@@ -192,6 +192,21 @@
+            ordinal = 1, 
+            target = "Lnet/minecraft/entity/item/EntityItem;setDead()V")
+    )
+    private void spongeImpl$fireExpireEntityEventTargetItem(final CallbackInfo ci) {
```
Not to be too nit picky:
`impl$fireExpireEntityEvent` is what the method name should be. Prefix is `impl$` not `spongeImpl$`

gabizou commented on src/main/java/org/spongepowered/common/mixin/core/entity/item/EntityItemMixin.java:
```diff
@@ -192,6 +192,21 @@
+        method = "onUpdate", 
+        at = @At (
+            value = "INVOKE", 
+            ordinal = 1, 
```
Could this be redone with a slice instead of a single invoke that looks a little vague without looking at the target method?

DrZoddiak replied:
So to be clear, I know nothing about mixins. My implementation was mostly based off of #1917 
and working around until it worked. That being said, I did give this a go trying to read a bit. 
(Honestly mixins aren't something I should be playing with)
I didn't want to upload it, as I have no idea if it is bad practice, but I did test the changes made here, and it still works.

https://gist.github.com/DrZoddiak/408a71583c974392beaf44d235c92c5d

gabizou commented on src/main/java/org/spongepowered/common/mixin/core/entity/item/EntityItemMixin.java:
```diff
@@ -192,6 +192,21 @@
+            target = "Lnet/minecraft/entity/item/EntityItem;setDead()V")
+    )
+    private void spongeImpl$fireExpireEntityEventTargetItem(final CallbackInfo ci) {
+        try (CauseStackManager.StackFrame frame = Sponge.getCauseStackManager().pushCauseFrame()) {
```
Is it guaranteed that we're on the main thread in single player? I remember at times there's a lot that is written assuming this is run on the server only to blow up on the client.

DrZoddiak replied:
I have no idea how to verify this without running the jar on my client, and I cannot seem to get my compiled jar to work on the client. As noted above I'm very inexperienced with this stuff.

gabizou reviewed (changes requested):
Quick look, address comments, need to test personally or another sponge dev to test.

DrZoddiak committed:
"nit pick" issue & add imports

DrZoddiak committed:
Use slice to be more specific

DrZoddiak committed:
Add check for Main thread

According to Doot from Discord
"Entities should all be singlethreaded"
I'm adding the check just in case a mod decides to spawn an entity Async

DrZoddiak committed:
Fix Missing Imports

DrZoddiak committed:
Fix issue with crashing

DrZoddiak commented:
For ease of testing I've made a small test plugin someone can just toss into their plugins folder.
Code:
https://gist.github.com/DrZoddiak/bab23726553ab28ae4194c2cf3ac7f2d
Jar:
https://github.com/DrZoddiak/SpongeCommon/releases/tag/1.0

gabizou commented on src/main/java/org/spongepowered/common/mixin/core/entity/item/EntityItemMixin.java:
```diff
@@ -29,14 +29,17 @@
-import org.spongepowered.asm.mixin.injection.Inject;
-import org.spongepowered.asm.mixin.injection.ModifyConstant;
-import org.spongepowered.asm.mixin.injection.Redirect;
+import org.spongepowered.asm.mixin.injection.*;
```
No star imports. Sorry for nitpicking.

gabizou commented on src/main/java/org/spongepowered/common/mixin/core/entity/item/EntityItemMixin.java:
```diff
@@ -192,6 +195,29 @@
+                ExpireEntityEvent.TargetItem event = SpongeEventFactory.createExpireEntityEventTargetItem(Sponge.getCauseStackManager().getCurrentCause(), (Item) this);
+                SpongeImpl.postEvent(event);
+            }
+        }
```
```suggestion
        if (!SpongeImplHooks.isMainThread() || this.getItem().isEmpty()) {
            // In the rare case the first if block is actually at the end of the method instruction list, we don't want to 
            // erroneously be calling this twice.
            return;
        }
        try (final CauseStackManager.StackFrame frame = Sponge.getCauseStackManager().pushCauseFrame()) {
            frame.pushCause(this);
            final ExpireEntityEvent.TargetItem event = SpongeEventFactory.createExpireEntityEventTargetItem(frame.getCurrentCause(), (Item) this);
            SpongeImpl.postEvent(event);
        }
```

gabizou commented on src/main/java/org/spongepowered/common/mixin/core/entity/item/EntityItemMixin.java:
```diff
@@ -192,6 +195,29 @@
     }
+    
+    @Inject(
+        method = "onUpdate", 
```
```suggestion
        method = "onUpdate",
        at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/item/EntityItem;setDead()V"),
        slice = @Slice(
            from = @At(value = "INVOKE", target = "Lnet/minecraft/entity/item/EntityItem;handleWaterMovement()Z"),
            to = @At("TAIL")
        )
```

gabizou reviewed (changes requested):
The injection point using ordinals is unclear, but what is clear is saying "target the method call `setDead()` after `handleWaterMovement()` is called" when you end up looking at the method. Because bytecode is weird and decompilation/recompilation is also weird, there's a corner case where the original if statement could be making a return at the end, where the `setDead()` is also after the instruction`handleWaterMovement()`. This ensures that the injection is going to consider that corner case where the event may be fired twice because the injection occurs twice.

DrZoddiak committed:
Remove wildcard import

DrZoddiak committed:
Update src/main/java/org/spongepowered/common/mixin/core/entity/item/EntityItemMixin.java

Co-Authored-By: Gabriel Harris-Rouquette <github@gabizou.com>

DrZoddiak committed:
Update src/main/java/org/spongepowered/common/mixin/core/entity/item/EntityItemMixin.java

Co-Authored-By: Gabriel Harris-Rouquette <github@gabizou.com>

DrZoddiak committed:
Fix commit issues

ImMorpheus reviewed (approved):
Working as intended
